src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +199,78 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
+            Socket resultSocket = null;
+            ExceptionDispatchInfo lastSocketConnectEx = null;
+
+            CancellationTokenSource cancelTokenSource = new CancellationTokenSource();
+            CancellationToken cancelToken = cancelTokenSource.Token;
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    sockets.Add(socket);
+                    Task socketTask = socket.ConnectAsync(address, port).ContinueWith(
+                        t =>
+                        {
+                            cancelToken.ThrowIfCancellationRequested();
+                            if (socket.Connected && !cancelToken.IsCancellationRequested)
+                            {
+                                lock (socketOpenTaskLock)
+                                {
+                                    if (resultSocket == null)
+                                    {
+                                        // Successful socket connection, so cancel other pending tasks
+                                        cancelTokenSource.Cancel();
+                                        resultSocket = socket;
+                                        sockets.Remove(resultSocket);
+
+                                        // Dispose any sockets that were still pending at time of success/error
+                                        foreach (Socket socketToDispose in sockets)
+                                        {
+                                            if (socketToDispose.Connected)
+                                            {
+                                                socketToDispose.Shutdown(SocketShutdown.Both);
+                                            }
+                                            socketToDispose.Dispose();
+                                        }
+                                    }
+                                }
+                            }
+                        },
+                        cancelToken);
+
+                    connectTasks.Add(socketTask);
                 }
                 catch (Exception exc)
                 {
                     socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    lastSocketConnectEx = ExceptionDispatchInfo.Capture(exc);
                 }
             }
 
-            // Propagate the last failure that occurred
-            if (lastException != null)
+            try
+            {
+                await Task.WhenAll(connectTasks).ConfigureAwait(false);
```

Comment:
Sockets that are still pending get cancelled when one socket successfully connects.